drivers/bms_ic/bq769x0/bq769x0.c

```diff
@@ -883,9 +883,39 @@ static int bms_ic_bq769x0_balance(const struct device *dev, uint32_t cells)
     return err;
 }
 
+static int bq769x0_activate(const struct device *dev)
+{
+    struct bms_ic_bq769x0_data *dev_data = dev->data;
+
+    int err = bq769x0_detect_crc(dev);
+    if (!err) {
+        /* switch external thermistor and ADC on */
+        bq769x0_write_byte(dev, BQ769X0_SYS_CTRL1, 0b00011000);
+        /* switch CC_EN on */
+        bq769x0_write_byte(dev, BQ769X0_SYS_CTRL2, 0b01000000);
+
+        /* get ADC offset (2's complement) and gain */
+        dev_data->adc_offset = (signed int)bq769x0_read_byte(dev, BQ769X0_ADCOFFSET);
+        dev_data->adc_gain = 365
+                             + (((bq769x0_read_byte(dev, BQ769X0_ADCGAIN1) & 0b00001100) << 1)
+                                | ((bq769x0_read_byte(dev, BQ769X0_ADCGAIN2) & 0b11100000) >> 5));
+    }
+    else {
+        LOG_ERR("BMS communication error");
+        return err;
+    }
+
+    /* run processing once at start-up to check and clear errors */
+    k_work_schedule(&dev_data->alert_work, K_NO_WAIT);
```

Comment:
Yes, it was intentional, as the alert work handler requests data from the device, so the CRC should be detected before that.

Actually, we should probably also put the callback initialization into this function. Otherwise the alert work handler could be fired before the activate function has been called. I'll fix the commit and add some more information to the commit message.

Some background information regarding the activation for you: We faced the issue that because of glitches in the I2C communication during start-up (see issue #25) the bq769x2 did not respond after start-up. However, a driver in Zephyr cannot be re-initialized once failed. For this reason we need the activation function, which can be called multiple times until it succeeds.

The activation can also be used for boards where the MCU is powered independently of the BMS IC. This is currently not the case for Libre Solar PCBs, but could be a useful addition for the future. Depending on the wakeup method for the chip, some additional GPIO definitions would probably have to be added to the chip's devicetree binding.

Reply:
Alert callback initialization moved as well.

Reply:
Thank you for the clarification and providing additional background information.